### PR TITLE
Charlesmchen/analytics3

### PR DESCRIPTION
**Pull request by charlesmchen**
A first pass at instrumenting some of the critical sections of our codebase.

PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
PTAL @michaelkirk

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 1]
```diff
@@ -42,6 +42,32 @@
...
+#define kOWSProdAssertParameterEnvelopeDescription @"envelope_description"
+#define kOWSProdAssertParameterEnvelopeEncryptedLength @"encrypted_length"
+
+#define AnalyticsParametersFromEnvelope(__envelope)                                                                    \
+    ^{                                                                                                                 \
+        NSData *__encryptedData = __envelope.hasContent ? __envelope.content : __envelope.legacyMessage;               \
+        return (@{                                                                                                     \
+            kOWSProdAssertParameterEnvelopeIsLegacy : @(__envelope.hasLegacyMessage),                                  \
+            kOWSProdAssertParameterEnvelopeHasContent : @(__envelope.hasContent),                                      \
+            kOWSProdAssertParameterEnvelopeDescription : [self descriptionForEnvelopeType:__envelope],                 \
```
Comment (michaelkirk) [inline SignalServiceKit/src/Messages/TSMessagesManager.m]: This reads a little goofy. Key is for `description` but value is the envelope type. Maybe the key should be `kOWSProdAssertParameterEnvelopeType`.

File: SignalServiceKit/src/Messages/TSMessagesManager.m [hunk 2]
```diff
@@ -367,15 +399,18 @@ - (void)handleSecureMessageAsync:(OWSSignalServiceProtosEnvelope *)messageEnvelo
...
                 return;
             }
 
             NSUInteger kMaxEncryptedDataLength = 250 * 1024;
             if (encryptedData.length > kMaxEncryptedDataLength) {
-                DDLogError(@"Skipping message envelope with oversize encrypted data: %lu.",
-                    (unsigned long)encryptedData.length);
+                OWSProdErrorWParams(@"message_manager_error_oversize_message", ^{
+                    return (@{
+                        @"message_size" : @(encryptedData.length),
```
Comment (michaelkirk) [inline SignalServiceKit/src/Messages/TSMessagesManager.m]: order of magnitude?

File: SignalServiceKit/src/Util/OWSAnalytics.h [hunk 3]
```diff
@@ -115,14 +117,33 @@ typedef NSDictionary<NSString *, id> *_Nonnull (^OWSProdAssertParametersBlock)()
...
+            kOWSProdAssertParameterNSExceptionReason : (__exception.reason ?: @"unknown"),                             \
+            kOWSProdAssertParameterNSExceptionClassName :                                                              \
+                (__exception ? NSStringFromClass([__exception class]) : @"unknown"),                                   \
         });                                                                                                            \
     }
 
 #define OWSProdFailWNSError(__analyticsEventName, __nserror)                                                           \
-    OWSProdFailWParams(__analyticsEventName, AnalyticsParametersFromNSError(__nserror))
+    {                                                                                                                  \
+        DDLogError(@"%s:%d %@: %@", __PRETTY_FUNCTION__, __LINE__, __analyticsEventName, __nserror.debugDescription);  \
```
Comment (michaelkirk) [inline SignalServiceKit/src/Util/OWSAnalytics.h]: Is this `DDLogError` call redundant?

below is OWProdFailWParams...

... calls `OWSProdFailWParamsTemplate`
... calls `OWSProdFailWParamsTemplate`
... calls `OWSAnalytics logEvent:`
... calls `LOG_MAYBE`

File: SignalServiceKit/src/Messages/TSMessagesManager.m (hunk 1 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/TSMessagesManager.m]: Okay.  I'm going to address this in the next PR where I've already been renaming these constants.

File: SignalServiceKit/src/Messages/TSMessagesManager.m (hunk 2 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/TSMessagesManager.m]: Done.

File: SignalServiceKit/src/Util/OWSAnalytics.h (hunk 3 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Util/OWSAnalytics.h]: We want the debug logs to be more verbose than the analytics events.  In this case the error's `debugDescription` is considered valuable enough to log but too dangerous to include in the analytics event.  I'll add a comment.

Comment (charlesmchen) [inline SignalServiceKit/src/Util/OWSAnalytics.h]: See: https://github.com/WhisperSystems/Signal-iOS/pull/2353

File: SignalServiceKit/src/Messages/TSMessagesManager.m (hunk 1 above)
Comment (charlesmchen) [inline SignalServiceKit/src/Messages/TSMessagesManager.m]: See: https://github.com/WhisperSystems/Signal-iOS/pull/2353

**Pull request merged**